read_header.cpp

```diff
@@ -87,12 +88,180 @@ int main() {
+                }
+                hydrogen_number_density[i][j][k] = rho * X / m_p; // Density * X / m_p
+                helium_number_density[i][j][k] = rho * X * He_to_H_ratio / m_p; // n_He/n_H=0.1
+
+            }
+        }
+    }
+    phys_var.push_back(Temperature); // Add the Temperature array to phys_var for further processing if needed
+    phys_var.push_back(electron_density); // Add the electron density array to phys_var for further processing if needed
+    phys_var.push_back(hydrogen_number_density); // Add the hydrogen number density array to phys_var for further processing if needed
+    phys_var.push_back(helium_number_density); // Add the helium number density array to phys_var for further processing if needed
+    std::cout << "Temperature calculated successfully and added to phys_var at index " << phys_var.size() - 2 << "\n";
+    std::cout << "Electron density calculated successfully and added to phys_var at index " << phys_var.size() - 1 << "\n";
+    std::cout << "Hydrogen number density calculated successfully and added to phys_var at index " << phys_var.size() << "\n";
+    std::cout << "Helium number density calculated successfully and added to phys_var at index " << phys_var.size() + 1 << "\n";
+    
+
+    // Calculate the density of ions: n_H+, n_He+, and n_He++
+    std::vector<std::vector<std::vector<double>>> H_plus_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::vector<std::vector<std::vector<double>>> He_plus_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::vector<std::vector<std::vector<double>>> He_double_plus_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));\
+    std::vector<std::vector<std::vector<double>>> Neutral_hydrogen_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::vector<std::vector<std::vector<double>>> Neutral_helium_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::cout << "Calculating ion densities...\n";
+    for (int i = 0; i < hinfo.global_nx; ++i) {
+        for (int j = 0; j < hinfo.global_ny; ++j) {
+            for (int k = 0; k < hinfo.global_nz; ++k) {
+                double n_e = electron_density[i][j][k];
+                double n_H = hydrogen_number_density[i][j][k];
+                double n_He = helium_number_density[i][j][k];
+                if(n_e < (n_H * (1 + He_to_H_ratio))) {
+                    // Assuming the ionization fraction, chi is same for both hydrogen and helium, we have n_e=n_H+ + n_He+  implying n_H+=n_H*n_e/(n_H+n_He)
+                    H_plus_density[i][j][k] = n_H * n_e / (n_H + n_He);
```

Comment:
The quantities that Criptic uses are xH0, xHp, xHe0, xHep, xHep2, xe which are abundances per H nucleon. So, we should be pushing those to physics vars.